packages/editor/src/components/local-autosave-monitor/index.js

```diff
@@ -0,0 +1,169 @@
+
+/**
+ * Internal dependencies
+ */
+import AutosaveMonitor from '../autosave-monitor';
+
+/**
+ * Time interval at which local autosave should occur after a change, in
+ * seconds.
+ *
+ * @var {number}
+ */
+const AUTOSAVE_INTERVAL_SECONDS = 15;
+
+/**
+ * Function which returns true if the current environment supports browser
+ * localStorage, or false otherwise. The result of this function is cached and
+ * reused in subsequent invocations.
+ */
+const hasLocalStorageSupport = once( () => {
+	try {
+		// Private Browsing in Safari 10 and earlier will throw an error when
+		// attempting to set into localStorage. The test here is intentional in
+		// causing a thrown error as condition bailing from local autosave.
+		window.localStorage.setItem( '__wpEditorTestLocalStorage', '' );
+		window.localStorage.removeItem( '__wpEditorTestLocalStorage' );
+		return true;
+	} catch ( error ) {
+		return false;
+	}
+} );
+
+/**
+ * Custom hook which returns a callback function to be invoked when a local
+ * autosave should occur.
+ *
+ * @return {Function} Callback function.
+ */
+function useAutosaveSaveCallback() {
+	const {
+		postId,
+		getEditedPostAttribute,
+	} = useSelect( ( select ) => ( {
+		postId: select( 'core/editor' ).getCurrentPostId(),
+		getEditedPostAttribute: select( 'core/editor' ).getEditedPostAttribute,
+	} ) );
+
+	return useCallback( () => {
+		window.localStorage.setItem( 'post_' + postId, JSON.stringify( {
+			post_title: getEditedPostAttribute( 'title' ),
+			content: getEditedPostAttribute( 'content' ),
+			excerpt: getEditedPostAttribute( 'excerpt' ),
+		} ) );
+	}, [ postId, getEditedPostAttribute ] );
+}
+
+/**
+ * Custom hook which manages the creation of a notice prompting the user to
+ * restore a local autosave, if one exists.
+ */
+function useAutosaveNotice() {
+	const {
+		postId,
+		getEditedPostAttribute,
+	} = useSelect( ( select ) => ( {
+		postId: select( 'core/editor' ).getCurrentPostId(),
+		getEditedPostAttribute: select( 'core/editor' ).getEditedPostAttribute,
+	} ) );
+
+	const { createWarningNotice, removeNotice } = useDispatch( 'core/notices' );
+	const { editPost, resetEditorBlocks } = useDispatch( 'core/editor' );
+
+	useEffect( () => {
+		let autosave = window.localStorage.getItem( 'post_' + postId );
+		if ( ! autosave ) {
+			return;
+		}
+
+		try {
+			autosave = JSON.parse( autosave );
+		} catch ( error ) {
+			// Not usable if it can't be parsed.
+			return;
+		}
+
+		const { post_title: title, content, excerpt } = autosave;
+		const edits = { title, content, excerpt };
+
+		// Only display a notice if there is a difference between what has been
+		// saved and that which is stored in localStorage.
+		const hasDifference = Object.keys( edits ).some( ( key ) => {
+			return edits[ key ] !== getEditedPostAttribute( key );
+		} );
+
+		if ( ! hasDifference ) {
+			// If there is no difference, it can be safely ejected from storage.
+			window.localStorage.removeItem( 'post_' + postId );
```

Comment:
This would mean that if a second user logs in to the same browser, they can restore an autosave made by the first user.

Not sure if that's an issue, but I know the current autosave system stores revisions by different users separately. One way to align with that but still keep the same localstorage key is to save the userId in the payload.

Reply:
> This key format would mean that if a second user logs in to the same browser, they can restore an autosave made by the first user.

Based on other comments, I think `sessionStorage` may be enough to resolve this.

Though technically it could still be an issue if the user logs out of the current user and into another user while still within the same tab / session. I'd be curious to know if / how this was addressed in the classic editor.

Reply:
Currently in core we clear autosave data from session storage when the user logs out: https://github.com/WordPress/wordpress-develop/blob/6dad32d2aed47e6c0cf2aee8410645f6d7aba6bd/src/wp-login.php#L103